MySQLdb/converters.py

```diff
@@ -66,7 +66,7 @@ def Unicode2Str(s, d):
     return s.encode()
 
 def Float2Str(o, d):
-    return '%.15g' % o
+    return '%.16e' % o
```

Comment:
Test fails with `AssertionError: Decimal('0.1') != 0.1`

See https://dev.mysql.com/doc/refman/5.7/en/number-literals.html.
> Approximate-value numeric literals are represented in scientific notation with a mantissa and exponent

Reply:
???

Reply:
Why the error happens when only changed how "float" converted?

Reply:
Before any changes:

```
In [57]: con = MySQLdb.connect()

In [58]: cursor = con.cursor()

In [59]: val = 0.1

In [60]: cursor.execute('SELECT %s', (val,))
Out[60]: 1

In [61]: result = cursor.fetchone()

In [62]: result
Out[62]: (Decimal('0.1'),)

In [63]: result == val
Out[63]: False
```

It happens because 
```
In [65]: MySQLdb.converters.Float2Str(0.1, None)
Out[65]: '0.1'
```

And MySQL treats `0.1` as `DECIMAL`, to make MySQL treat it as 'REAL' it must be formatted like this (see https://dev.mysql.com/doc/refman/5.7/en/number-literals.html):

```
In [70]: cursor.execute('SELECT 0.1e0')
Out[70]: 1

In [71]: cursor.fetchone()[0]
Out[71]: 0.1
```

Reply:
I don't think it's problem.
I prefer `3.14` to `3.14000000000000e+00` when sending float 3.14.

Reply:
There is an edge case that I just recently discovered when trying to insert 1.7976931348623157e+308 (max double), the `'%.15g'` rounds it to '1.79769313486232e+308' and `'%.16g'` rounds to '1.797693134862316e+308' which are both ever so slightly larger than 1.7976931348623157e+308, and throws an error.

Reply:
Doesn't repr() solve it?

Reply:
> I don't think it's problem.
> I prefer `3.14` to `3.14000000000000e+00` when sending float 3.14.

@methane 
I think it's a problem. I don't see any good reason to send floats as `DECIMAL` literals.
Why when I send `float(0.1)` I get back `Decimal('0.1')`? And when I send `float(0.00001)` I get back `1e-05`. That's odd.

For example I have a function:
```
def func(cursor, v):
    cursor.execute('SELECT f * %s FROM t', (v,))  # f is INT field
    return timedelta(0, 0, cursor.fetchone()[0])
```
I expect `func(cursor, 0.1)` to work since MySQL return float for int * float, but instead I have `TypeError: unsupported type for timedelta microseconds component: decimal.Decimal`.

Also it can affect performance:
```
mysql> SELECT COUNT(*) FROM t;
+----------+
| COUNT(*) |
+----------+
| 20000000 |
+----------+
1 row in set (4.78 sec)

mysql> SELECT SUM(f * 0.1e0) FROM t;
+--------------------+
| SUM(f * 0.1e0)     |
+--------------------+
| 10009115659.990402 |
+--------------------+
1 row in set (5.26 sec)

mysql> SELECT SUM(f * 0.1) FROM t;
+---------------+
| SUM(f * 0.1)  |
+---------------+
| 10009115660.0 |
+---------------+
1 row in set (6.52 sec)
```

Reply:
> How about `s = repr(d); if 'e' not in s: s += 'e0'`?
> 
> '3.14e0' seems much better than '3.1400000000000001e+00'

Yes, it's obviously better, I'll fix it.

> Column type is important, but literal type is not important for this project.

I just want to clarify what you mean. Consider this example related to #267:
```
cursor.execute('CREATE TABLE t (f DECIMAL(65,30))')
cursor.execute('INSERT INTO t(f) VALUES (%s)', (Decimal('0.1'),))
cursor.execute('SELECT f FROM t WHERE f + %s = 0.3', (Decimal('0.2'),))
```
If `Decimal('0.2')` is formatted as `0.2` (decimal) 1 row is returned, if it's formatted as `'0.2'`(current behavior, MySQL treats it as float) 0 rows are returned. Considering this, is literal type important for this project?

Reply:
> I just want to clarify what you mean. Consider this example related to #267:
> 
> ```
> cursor.execute('CREATE TABLE t (f DECIMAL(65,30))')
> cursor.execute('INSERT INTO t(f) VALUES (%s)', (Decimal('0.1'),))
> cursor.execute('SELECT f FROM t WHERE f + %s = 0.3', (Decimal('0.2'),))
> ```
> If `Decimal('0.2')` is formatted as `0.2` (decimal) 1 row is returned, if it's formatted as `'0.2'`(current behavior, MySQL treats it as float) 0 rows are returned. Considering this, is literal type important for this project?

It's totally different thing.  Sending number in string format is considered a bug, of course.
On the other hand, using normal numeric like `3.14` which MySQL considers Decimal when inserting to float column is not a bug.  It's usually used, and not considered bug because Decimal -> float conversion is safe.